### PR TITLE
WIP : Storing processing logs and failures in excel file

### PR DESCRIPTION
**Pull request by agramfort**
@hoechenberger I would love your input on this one

### CODE REVIEW

**Comment by hoechenberger** [comment]
@agramfort
Seems the JSON dump fails if `mri_t1_path_generator` is set (function cannot be converted to JSON). I would say that the solution to that is to replace the value of `mri_t1_path_generator` with a string such as `<func: function_name>` before converting to JSON. WDYT?

**Comment by hoechenberger** [comment]
(Alternatively, we could dump the entire function content … But that might be too much?)

File: config.py [hunk 1]
```diff
@@ -2824,6 +2841,40 @@ def get_eeg_reference() -> Union[Literal['average'], Iterable[str]]:
...
+        writer.book = book
+    else:
+        writer = pd.ExcelWriter(fname, engine='openpyxl')
+
+    df.to_excel(writer, sheet_name=sheet_name, index=False)
+    writer.save()
+    writer.close()
+
+
+
```
Comment (hoechenberger) [inline config.py:2877]:
```suggestion


```

File: config.py [hunk 2]
```diff
@@ -2078,47 +2122,86 @@ def get_fs_subjects_dir():
...
 
 
 def failsafe_run(on_error):
     def failsafe_run_decorator(func):
         @functools.wraps(func)  # Preserve "identity" of original function
         def wrapper(*args, **kwargs):
+            kwargs_copy = copy.deepcopy(kwargs)
+            if "cfg" in kwargs_copy:
+                kwargs_copy["cfg"] = json_tricks.dumps(
+                    kwargs_copy["cfg"], sort_keys=False, indent=2
```
Comment (hoechenberger) [inline config.py:2179]:
```suggestion
                    kwargs_copy["cfg"], sort_keys=False, indent=4
```

File: config.py [hunk 3]
```diff
@@ -2078,47 +2122,86 @@ def get_fs_subjects_dir():
...
+                )
+            log_info = pd.Series(kwargs_copy)
             try:
-                return func(*args, **kwargs)
+                assert len(args) == 0  # make sure params are only kwargs
+                out = func(*args, **kwargs)
+                assert out is None  # make sure the function return None
```
Comment (hoechenberger) [inline config.py:2185]:
```suggestion
                assert out is None  # make sure the function returned None
```

File: config.py [hunk 4]
```diff
@@ -2078,47 +2122,86 @@ def get_fs_subjects_dir():
...
-    if prefix:
-        prefix = f'[{prefix}]'
+    message = f' {message}'
 
-    if step is not None:
-        prefix = f'[Step-{step:02}]{prefix}'
+    # prefix = ', '.join([item for item in [subject, session, run]
+    #                     if item is not None])
+    # if prefix:
+    #     prefix = f'[{prefix}]'
```
Comment (hoechenberger) [inline config.py:2148]:
```suggestion
```

File: config.py [hunk 5]
```diff
@@ -1531,10 +1533,52 @@ def get_t1_from_meeg(bids_path):
...
 logger = logging.getLogger('mne-bids-pipeline')
 
-log_fmt = '%(asctime)s %(message)s'
+log_fmt = '[%(asctime)s] %(step)s%(subject)s%(session)s%(run)s%(message)s'
 log_date_fmt = coloredlogs.DEFAULT_DATE_FORMAT = '%H:%M:%S'
-coloredlogs.install(level=log_level, logger=logger, fmt=log_fmt,
-                    date_fmt=log_date_fmt)
+log_level_styles = {
+    'warning': '202,bold;error=background=red',
+    'critical': 'bold,background=red'
```
Comment (hoechenberger) [inline config.py:1540]:
```suggestion
    'warning': '202,bold',
    'error': 'bold,background=red',
    'critical': 'bold,background=red'
```

File: config.py [hunk 6]
```diff
@@ -2824,6 +2903,39 @@ def get_eeg_reference() -> Union[Literal['average'], Iterable[str]]:
...
+    sheet_name = sheet_name[-30:]  # shorten due to limit of excel format
+
+    df = pd.DataFrame(logs)
+
+    columns = df.columns
+    if "cfg" in columns:
+        columns = list(columns)
+        idx = columns.index("cfg")
+        del columns[idx]
+        columns.insert(-2, "cfg")
```
Comment (hoechenberger) [inline config.py:2922]: How about sth like:

```python
columns = ['subject', 'session', 'run', 'task', 'cfg', 'error']

for idx, column in enumerate(columns.copy()):
    if column not in df.columns:
        del columns[idx]
```
Wouldn't that give us more control?

**Review by hoechenberger — APPROVED** [review]

File: config.py [hunk 7]
```diff
@@ -1531,13 +1533,109 @@ def get_t1_from_meeg(bids_path):
...
+coloredlogs.install(
+    level=log_level, logger=logger, fmt=log_fmt,
+    date_fmt=log_date_fmt, level_styles=log_level_styles,
+    field_styles=log_field_styles
+)
+
 
 mne.set_log_level(verbose=mne_log_level.upper())
 
+class LogKwargsT(TypedDict):
```
Comment (hoechenberger) [inline config.py:1599]:
```suggestion

class LogKwargsT(TypedDict):
```

**Review by hoechenberger — APPROVED** [review]

**Comment by agramfort** [comment]
just tested with @apmellot and it works like a charm !

**Comment by hoechenberger** [comment]
amazing!!!!!! 🥳 🍾

**Comment by apmellot** [comment]
Thank you very much @hoechenberger and @agramfort, this will change my life ! :smile:

**Comment by larsoner** [comment]
Looks like this PR broke builds on `main`, see failures in #435 and #433

**Pull request merged**